hack/e2e.go

```diff
@@ -89,57 +94,129 @@ func main() {
 		cmd.Stdin = os.Stdin
 		if !finishRunning("build-release", cmd) {
 			log.Fatal("Error building. Aborting.")
 		}
 	}
 
-	os.Setenv("KUBECTL", *root+`/cluster/kubectl.sh`+kubectlArgs())
+	if *up && !TearDown() {
+		log.Fatal("Could not tear down previous cluster")
+	}
 
-	if *pushup {
-		if IsUp() {
-			log.Printf("e2e cluster is up, pushing.")
-			*up = false
-			*push = true
-		} else {
-			log.Printf("e2e cluster is down, creating.")
-			*up = true
-			*push = false
-		}
+	beforeResources := ""
+	if *checkLeakedResources {
+		beforeResources = ListResources()
 	}
-	if *up {
-		if !Up() {
-			log.Fatal("Error starting e2e cluster. Aborting.")
+
+	os.Setenv("KUBECTL", strings.Join(append([]string{"./cluster/kubectl.sh"}, kubectlArgs()...), " "))
+
+	if *upgrade_args != "" { // Start the cluster using a previous version.
+		if !UpgradeUp() {
+			log.Fatal("Failed to start cluster to upgrade. Aborting.")
+		}
+	} else { // Start the cluster using this version.
+		if *pushup {
+			if IsUp() {
+				log.Printf("e2e cluster is up, pushing.")
+				*up = false
+				*push = true
+			} else {
+				log.Printf("e2e cluster is down, creating.")
+				*up = true
+				*push = false
+			}
 		}
-	} else if *push {
-		if !finishRunning("push", exec.Command(path.Join(*root, "hack/e2e-internal/e2e-push.sh"))) {
-			log.Fatal("Error pushing e2e cluster. Aborting.")
+		if *up {
+			if !Up() {
+				log.Fatal("Error starting e2e cluster. Aborting.")
+			}
+		} else if *push {
+			if !finishRunning("push", exec.Command("./hack/e2e-internal/e2e-push.sh")) {
+				log.Fatal("Error pushing e2e cluster. Aborting.")
+			}
 		}
 	}
 
 	success := true
-	switch {
-	case *ctlCmd != "":
+
+	if *ctlCmd != "" {
 		ctlArgs := strings.Fields(*ctlCmd)
 		os.Setenv("KUBE_CONFIG_FILE", "config-test.sh")
-		success = finishRunning("'kubectl "+*ctlCmd+"'", exec.Command(path.Join(*root, "cluster/kubectl.sh"), ctlArgs...))
-	case *test:
-		success = Test()
+		success = success && finishRunning("'kubectl "+*ctlCmd+"'", exec.Command("./cluster/kubectl.sh", ctlArgs...))
+	}
+
+	if *upgrade_args != "" {
+		success = success && UpgradeTest(*upgrade_args)
+	}
+
+	if *test {
+		if *skewTests {
+			success = success && SkewTest()
+		} else {
+			success = success && Test()
+		}
+	}
+
+	if *kubemark {
+		success = success && KubemarkTest()
 	}
 
 	if *down {
-		TearDown()
+		if !success && *dump != "" {
+			dumpClusterLogs(*dump)
+		}
+		success = success && TearDown()
+	}
+
+	if *checkLeakedResources {
+		DiffResources(beforeResources, ListResources())
 	}
 
 	if !success {
 		os.Exit(1)
 	}
 }
 
+func DiffResources(before string, after string) {
+	differ := diffmatchpatch.New()
+	problem := 0
+	for _, diff := range differ.DiffMain(before, after, true) {
+		switch diff.Type {
+		case diffmatchpatch.DiffInsert:
+			problem++
+			for _, line := range strings.Split(diff.Text, "\n") {
+				log.Printf("+ %s", line)
+			}
+		case diffmatchpatch.DiffDelete:
```

Comment:
Accounting deletion caused false positives in the past. Scenario is following:
- run X fails to delete resource A
- run X+1 calls kube-down, deletes resource A, finishes and cleans up after itself and fails because we're missing A.